### ISSUE TITLE
5.0 Hosted Blazor WASM CORS error

### ISSUE THREAD

**Issue by adds39939** [dotnet/AspNetCore.Docs#20482]
[EDIT by guardrex to add the metadata]

After upgrading from .NET Core 3.1 to .NET 5.0 I cannot login to my application successfully using Azure AD. Other than changing the project target sdks, the code has not been changed since .NET Core 3.1.

The following is displayed in the browser console after logging in once redirected back to the app:

```
Access to fetch at 'https://login.microsoftonline.com/[X]/oauth2/v2.0/token' from origin 'https://localhost:44396' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

AuthenticationService.js:32 
POST https://login.microsoftonline.com/[X]/oauth2/v2.0/token net::ERR_FAILED
```
Client ID is replaced with [X]

The wasm app gets stuck on `Completing login...` when this happens.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 3039ff79-ee86-20b1-0082-d3899a84c823
* Version Independent ID: af4a3a75-ae52-5f6d-2120-737bc2fefb07
* Content: [Secure an ASP.NET Core Blazor WebAssembly hosted app with Azure Active Directory](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory?view=aspnetcore-5.0)
* Content Source: [aspnetcore/blazor/security/webassembly/hosted-with-azure-active-directory.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/security/webassembly/hosted-with-azure-active-directory.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Hello @adds39939 ... It's best to open issues from the bottom of the topic that you're following. No worries tho ... I'll add the metadata manually to the issue.

Are you working with a standalone AAD app or a hosted AAD app?

WRT your app migration, the migration guidance is here ...

https://docs.microsoft.com/aspnet/core/migration/31-to-50?view=aspnetcore-5.0&tabs=visual-studio#update-blazor-webassembly-projects

... and the topics have been updated for 5.0 ...

* Standalone: https://docs.microsoft.com/aspnet/core/blazor/security/webassembly/standalone-with-azure-active-directory?view=aspnetcore-5.0
* Hosted: https://docs.microsoft.com/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory?view=aspnetcore-5.0

**Comment by adds39939** [comment]
Hi there,

Apologies about the placement of the issue, I'll be more careful next time.

The app is ASP.NET Core Hosted Blazor Wasm. I have followed the migration guidance as stated above, however, placing the 3.1 app and the 5.0 app side by side presents the issue stated.

**Comment by guardrex** [comment]
>  side by side

That seems a little scary. 🎃

Lingering cookies across sessions while testing can break these apps fairly easily. Are you using a new in-private/incognito browser for **_every test run_** no matter how small the change (either app change or Azure portal change)? It's important to rule out lingering cookie problems. If you look at the *Troubleshoot* section of any of the Blazor WASM security topics, there's guidance there on configuring VS to do it automatically for you while testing.

**Comment by adds39939** [comment]
Sorry, I should have been more clear. After running the apps in succession, clearling all cache/data/cookies, I get the issue. Tested across Chrome, Firefox and Edge.

**Comment by guardrex** [comment]
What about that auto-generated port for IIS Express? Did you have that registered in the callback address in the Azure portal for this app (`44396`)?

Also note that ...

> clearling all cache/data/cookies

... is hit or miss. I know it seems like you'd be able to get it clean like that; but imo after working with these scenarios quite a bit, you're far better off with a fresh in-private/incognito browser **_for every test run_** no matter how small the change on either side. It always yields a nice, clean testing environment while troubleshooting.

Oh ... and you just reminded me of something that I need to add to the migration topic that I forgot about yesterday: The packages are different now. It's covered in the WASM topics, but the migration topic needs a touch for the *Sever* app of these hosted scnearios ...

```xml
<PackageReference Include="Microsoft.Identity.Web" Version="{VERSION}" />
<PackageReference Include="Microsoft.Identity.Web.UI" Version="{VERSION}" />
```

I'll do that right now. 🏃

**Comment by adds39939** [comment]
Yes I have tried this prior to submitting the issue. I have configured the Server and Client with the new MS Identity packages and ensured that the login callback address matches that of IIS. The issue still remains the same.

**Comment by guardrex** [comment]
Let me dig around a bit and get back to you. I didn't see that error here during migration, but I may have seen it in a product unit issue. I'll get back to you shortly.

**Comment by adds39939** [comment]
Thanks 😊

**Comment by guardrex** [comment]
Did you change over to the **SPA platform config** in the Azure portal for this app, or is it still **Web**?

**Comment by adds39939** [comment]
I have tried both, it was originally web so I changed to to SPA and the issue still persisted.

**Comment by guardrex** [comment]
I see that when we did the updates to the migration topic, we only indicated the change for standalone, but the correct platform config for standalone **_and the client app of a hosted solution_** is **SPA**. I'm fixing that now.

**Comment by guardrex** [comment]
What's the scheme of your App ID URI? ... is it `https://` like this ...

```
https://contoso.onmicrosoft.com/41451fa7-82d9-4673-8fa5-69eff5a761fd/API.Access
```

... or `api://` like this ...

```
api://41451fa7-82d9-4673-8fa5-69eff5a761fd/API.Access
```

**Comment by adds39939** [comment]
I have just created a new project using [this](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory?view=aspnetcore-5.0) template and I'm facing the same issue. [Here is a screenshot](https://i.imgur.com/hPtkHAC.png) of what happens after I login. Please note I can definitely login, the issue happens after the login-callback. I also have my Azure portal configured like [this](https://i.imgur.com/RBOQlrq.png).

**Comment by adds39939** [comment]
I have also tried the following App ID URI forms:
```
api://41451fa7-82d9-4673-8fa5-69eff5a761fd/API.Access
41451fa7-82d9-4673-8fa5-69eff5a761fd/API.Access
```

**Comment by guardrex** [comment]
You might be hitting this ...

https://docs.microsoft.com/en-us/aspnet/core/migration/31-to-50?view=aspnetcore-3.1&tabs=visual-studio#unauthorized-client-for-azure-active-directory-aad

... check the app's manifest.

**Comment by guardrex** [comment]
btw --

> api://

... that's good. You don't need the `Authority` with that type of App ID URI. :+1:

**Comment by guardrex** [comment]
... and disable implicit grant.

Also, that redirect uri isn't right ... try ...

```
https://localhost:44327/authentication/login-callback
```

**Comment by guardrex** [comment]
Ok ... I merged all of the updates and triggered a merge to the live branch. I was originally going to rely on cross-links to updated guidance sections in the topics, but there's just no way that would have worked. The info is scattered in a sea of complex content. I just placed all of the security updates into the migration topic ..... *I think* 🤞 ... there are **_a lot_** of updates. Those changes will go live in a few minutes.

**Comment by adds39939** [comment]
`allowPublicClient` in the manifest was already set to `null`, I changed it to `true` but nothing has changed.

I have changed the redirect URI to what you advised, still nothing. And implicit grant is disabled.

**Comment by guardrex** [comment]
Ok ... it's live here now. That was **_FAST!_** 🏃

https://docs.microsoft.com/en-us/aspnet/core/migration/31-to-50?view=aspnetcore-3.1&tabs=visual-studio#update-blazor-webassembly-projects

**Comment by guardrex** [comment]
Hummmmmm ... well ... we're running out of options before having to open a formal engineering issue.

Let's clean everything up and try one more time before opening an issue with them ...

Bit of a longshot, but when you went with 5.0 GA today ... the new SDK and updated your package references in the apps of the hosted solution, did you do the normal ✨ **_magic things_** ✨ to ward off evil spiritis 😈? ... Try this ... probably won't work ... but let's just make sure .................

* Delete the `obj` and `bin` folders of the projects in the solution (Client, Server, and Shared projects).
* Nuke the package caches by executing `dotnet nuget locals --clear all` in a command shell.
* Restore the solution ... compile and run it.

**Comment by guardrex** [comment]
Oh ... and you are on VS 16.8, right?

**Comment by adds39939** [comment]
Yeah that's right. Latest version of VS running .NET 5.0 released today. I have deleted the `obj` and `bin` folders in the template project and removed the nuget package caches. Same issue 😞. I do appreciate your help though 👍

**Comment by guardrex** [comment]
Ugh ... well ... that's about all that I can think of.

Open a new issue for the product unit at ...

https://github.com/dotnet/aspnetcore/issues

... give them the basic rundown of what you did for the upgrade. Show the relevant parts of the app ... best to include code snippets of the Client `Program.Main` and Server `Startup.ConfigureServices` with the `appsettings.json` file (of course replacing sensitive info with placeholders), show the screenshot of the error in the dev tools console, and the screenshot of the Azure portal config. Note them that you did check on the `allowPublicClient` manifest setting.

Sorry we didn't reach a happy end here, but this could be a real 🐞. They'll help figure out what the problem is.

Leave this issue open for the time being. I'll close it if it turns out to be a framework 🐞. I'll use it to track docs work if we need to add something to the migration topic.

For your new engineering issue, please add a `cc: @guardrex` to your opening remark so that I can follow along.

**Comment by guardrex** [comment]
HOLD a sec ... did you say "ClientId" in the `Authority`. That should be the **TenantId** there. Check that and confirm.

**Comment by guardrex** [comment]
Example ...

```json
{
  "AzureAd": {
    "Authority": "https://login.microsoftonline.com/{TENANT ID}",
    "ClientId": "{CLIENT APP CLIENT ID}",
    "ValidateAuthority": true
  }
}
```

```json
{
  "AzureAd": {
    "Authority": "https://login.microsoftonline.com/e86c78e2-...-918e0565a45e",
    "ClientId": "4369008b-21fa-427c-abaa-9b53bf58e538",
    "ValidateAuthority": true
  }
}
```
}

**Comment by adds39939** [comment]
```{
  "AzureAd": {
    "Authority": "https://login.microsoftonline.com/e86c78e2-...-918e0565a45e",
    "ClientId": "4369008b-21fa-427c-abaa-9b53bf58e538",
    "ValidateAuthority": true
  }
}
```

Yeah that's definitely how mine is configured.

**Comment by guardrex** [comment]
Ok ... it caught my :eye: when you had ...

> Client ID is replaced with [X]

.... up there. You merely meant "Tenant ID" when you typed that. Ok, cool.

**Comment by adds39939** [comment]
Ok this is interesting. I re-migrated the application from 3.1 to 5.0 on a different machine and the login works. I cloned the code over to the original machine and I have the CORS issue again. I guess you could say this is resolved? Though I have no clue on why it isn't working on the original machine despite everything being fine on 3.1.

**Comment by guardrex** [comment]
Is there a difference in OS and/or browser versions between the two machines?

**Comment by adds39939** [comment]
Both on Windows 10, the first is a work laptop and second is personal. Perhaps there is a policy or something on the work machine. Can't imagine why it would affect 3.1 -> 5.0 however.

**Comment by guardrex** [comment]
You could open that for engineering if you want. It would be a heck of a lot faster and easier to open. You just need to tell them that situation with an upgraded hosted Blazor security app and show the dev tools console screenshot. I think you should ask them; but if you don't, I'll keep an :ear: open for similar reports as we go forward with 5.0.

**Comment by adds39939** [comment]
Yeah I will do absolutely. Thanks for your help anyways!

**Comment by guardrex** [comment]
Ok, cool.

Please add a `cc: @guardrex` to it.

Leave this open. We might need to call something out depending on what they say.

**Comment by morganc9** [comment]
Hello,
I just did the migration from 3.1 to .Net 5. I have the same issue but with Azure B2c. Login page prompt and then after redirect I am still on "Completing login...". But I'm never properly login (even if I refresh...)
So I recreated a project from scratch by using the command line is here : https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-5.0
And copy pasting my confs... Still the same :/

**Comment by guardrex** [comment]
@morganc9 It will require investigation by engineering. When @adds39939 opens the issue over there, you can add a :+1: to it and/or a comment if there are any deltas between you or additional info that you'd like to add.

**Comment by morganc9** [comment]
Hello,
I just did the migration from 3.1 to .Net 5. I have the same issue but with Azure B2c. Login page prompt and then after redirect I am still on "Completing login...". But I'm never properly login (even if I refresh...)
So I recreated a project from scratch by using the command line is here : https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-5.0
And copy pasting my confs... Still the same :/

**Comment by morganc9** [comment]
Mhhh I fixed my issue... Not sure we should keep my post then.
I followed that: https://docs.microsoft.com/en-gb/azure/active-directory/develop/scenario-spa-app-registration#create-the-app-registration
You have to go to the portal and upgraded the uri of my UI + disabled the grant types.
sorry for the inconvenience :/

**Comment by guardrex** [comment]
No worries @morganc9 ... glad to hear it worked out. :tada:

**Comment by guardrex** [comment]
I'm going to close given this is a **_support request_** at this point. @adds39939 ... If you get an answer that leads to an apparent need to call sometime out in the topic, either post it on this issue and I'll re-open this or open a new issue for triage here.
